{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"16vzIDl3KrXEbXiAvNL5h4Pf6WdUtdonn","timestamp":1701355096452},{"file_id":"17xz1p8tzdaHJjjMwHMtzx11eDGwwE37d","timestamp":1696943812874}]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["Variable instructions:\n","\n","\n","\n","*   \"mutations_file_name\": name of the file stored in MyDrive\n","*   \"plasmid\": capitalised nucleotides for the entire plasmid sequence\n","*   \"insert\": capitalised nucleotides for the entire insert sequence (including STOP codon)\n","*   \"desired_tm\": integer value for the desired Tm across all primers\n","*   \"tm_range\": integer value for the accepted range across all primers (note: 1.2 has worked well previously)\n","*   \"lib_name\": library name to be generated using the primers (in 'LIBxxx' format)"],"metadata":{"id":"JoS86jQ2HtXA"}},{"cell_type":"code","source":["#@title Input variables, then hit 'Runtime' -> 'Run all'\n","mutations_file_name = \"FN1_SDM.csv\" # @param {type:\"string\"}\n","plasmid = \"\" # @param {type:\"string\"}\n","insert = \"\" # @param {type:\"string\"}\n","desired_tm = 72 # @param {type:\"integer\"}\n","tm_range = \"1.2\" # @param {type:\"string\"}\n","lib_name = \"FN1821\" # @param {type:\"string\"}"],"metadata":{"id":"p3Z_pvUC0O23","executionInfo":{"status":"ok","timestamp":1701362160049,"user_tz":0,"elapsed":9,"user":{"displayName":"Henry Taunt","userId":"07657040803472197154"}},"cellView":"form"},"execution_count":1,"outputs":[]},{"cell_type":"markdown","source":["# The code (click toggle to view) generates the single point mutants and stores an output Excel file on the common folder. Do not edit the code."],"metadata":{"id":"KGvFBGry9eQO"}},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive') # Mount gdrive"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"FtsfrP3v34SC","outputId":"9d341c50-26ce-4076-b14c-04d36579641e","executionInfo":{"status":"ok","timestamp":1701362162287,"user_tz":0,"elapsed":2246,"user":{"displayName":"Henry Taunt","userId":"07657040803472197154"}}},"execution_count":2,"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"]}]},{"cell_type":"code","source":["# Installation of packages and libraries\n","\n","!pip install biopython # Installs Biopython\n","from Bio.SeqUtils import MeltingTemp as mt # Imports the Tm calculator\n","from Bio.SeqUtils import Seq # Imports the Seq class\n","from Bio.SeqUtils import gc_fraction # Imports the GC fraction calculator\n","import pandas as pd"],"metadata":{"id":"z8OIWrjg5gL1","executionInfo":{"status":"ok","timestamp":1701362173953,"user_tz":0,"elapsed":11674,"user":{"displayName":"Henry Taunt","userId":"07657040803472197154"}},"colab":{"base_uri":"https://localhost:8080/"},"outputId":"2a8372cd-3da6-4383-b8f5-32181ba3bb03"},"execution_count":3,"outputs":[{"output_type":"stream","name":"stdout","text":["Requirement already satisfied: biopython in /usr/local/lib/python3.10/dist-packages (1.81)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.23.5)\n"]}]},{"cell_type":"code","source":["# Open the mutations and store them in a list\n","\n","mutations = open('/content/drive/My Drive/' + mutations_file_name, 'r', encoding='utf-8-sig')\n","\n","submitted_mutations = [] # Create a list to store the mutants\n","\n","for line in mutations:\n","  line = line.strip() # Remove empty lines from the mutation list\n","  submitted_mutations.append(line)"],"metadata":{"id":"Hsze2otH37dF","executionInfo":{"status":"ok","timestamp":1701362174296,"user_tz":0,"elapsed":345,"user":{"displayName":"Henry Taunt","userId":"07657040803472197154"}}},"execution_count":4,"outputs":[]},{"cell_type":"code","source":["# Generate dictionary of most frequently used codons for each amino acid in E. coli\n","\n","ecoli_codons = {'G':'GGC',\n","                'E':'GAA',\n","                'D':'GAT',\n","                'V':'GTT',\n","                'A':'GCG',\n","                'R':'CGC',\n","                'S':'AGC',\n","                'K':'AAA',\n","                'N':'AAC',\n","                'M':'ATG',\n","                'I':'ATT',\n","                'T':'ACC',\n","                'W':'TGG',\n","                'C':'TGC',\n","                'Y':'TAT',\n","                'L':'CTG',\n","                'F':'TTT',\n","                'Q':'CAG',\n","                'H':'CAT',\n","                'P':'CCG'}"],"metadata":{"id":"38K3PNS7bPU6","executionInfo":{"status":"ok","timestamp":1701362174296,"user_tz":0,"elapsed":3,"user":{"displayName":"Henry Taunt","userId":"07657040803472197154"}}},"execution_count":5,"outputs":[]},{"cell_type":"code","source":["# Store the provided form data as varialbes\n","\n","plasmid = plasmid\n","insert = insert\n","desired_tm = desired_tm\n","upper_tm = desired_tm + float(tm_range)\n","lower_tm = desired_tm - float(tm_range)"],"metadata":{"id":"ktjWxF567ue4","executionInfo":{"status":"ok","timestamp":1701362174296,"user_tz":0,"elapsed":3,"user":{"displayName":"Henry Taunt","userId":"07657040803472197154"}}},"execution_count":6,"outputs":[]},{"cell_type":"code","source":["# Finds the index where the insert starts in the plasmid\n","\n","start_index = plasmid.find(insert)"],"metadata":{"id":"I56j4V9mN6c3","executionInfo":{"status":"ok","timestamp":1701362174297,"user_tz":0,"elapsed":3,"user":{"displayName":"Henry Taunt","userId":"07657040803472197154"}}},"execution_count":7,"outputs":[]},{"cell_type":"code","source":["# This cell generates Fwd and Rev primers, ensuring they have the correct Tm and GC clamp properties\n","\n","fwd_primers = {} # Create a dictionary to store the Fwd primers\n","rev_primers = {} # Create a dictionary to store the Rev primers\n","\n","# Generate a Fwd primer with a GC clamp and calculate the Tm\n","for mutant in submitted_mutations:\n","  n = 1\n","  aa_to_introduce = mutant[len(mutant) - 1] # Identifies the AA to introduce\n","  index = mutant[1:len(mutant) - 1] # Identifies the index to mutate\n","  mut_index_in_bp = (int(index) * 3) + start_index # Identifies the index to mutate in bp and adds it to the insert start site index in the plasmid\n","  fwd_primer = ecoli_codons[aa_to_introduce] + plasmid[mut_index_in_bp:mut_index_in_bp + 30] # Adds the new AA as the first AA and appends 35 bp immediately downstream in the plasmid sequence\n","  #while 'G' not in fwd_primer[-1] and 'C' not in fwd_primer[-1]:\n","  #  fwd_primer = ecoli_codons[aa_to_introduce] + plasmid[mut_index_in_bp:mut_index_in_bp + 30 + n] # Lengthens the Fwd primer by one bp at a time if there is no GC clamp at the 3' position\n","  #  n += 1\n","  fwd_primer_tm = mt.Tm_NN(fwd_primer, nn_table=mt.DNA_NN3)\n","\n","  # Checks whether the Fwd primer Tm is too high - if so, it shortens the primer until the correct Tm is found\n","  k = 0\n","  while fwd_primer_tm > upper_tm:\n","    fwd_primer = fwd_primer[:len(fwd_primer) - 1] # Shortens the Fwd primer by one bp at a time if the\n","    #if 'G' in fwd_primer[-1] or 'C' in fwd_primer[-1]: # Only calculates a new Fwd primer Tm if it has a GC clamp\n","    fwd_primer_tm = mt.Tm_NN(fwd_primer, nn_table=mt.DNA_NN3)\n","    k += 1\n","    if k > 15:\n","      #continue # Ensures that an infinite loop (where Tm and GC clamp requirements cannot be matched, e.g. if A/T rich region) is broken\n","      break\n","  # Checks whether the Fwd primer Tm is too low\n","  n = 1\n","  k = 0\n","  while fwd_primer_tm < lower_tm:\n","    fwd_primer = ecoli_codons[aa_to_introduce] + plasmid[mut_index_in_bp:mut_index_in_bp + 35 + n] # add a character to increase the tm\n","    n += 1\n","    #print(fwd_primer,n,k,fwd_primer_tm)\n","    #if 'G' in fwd_primer[-1] or 'C' in fwd_primer[-1]:\n","    fwd_primer_tm = mt.Tm_NN(fwd_primer, nn_table=mt.DNA_NN3)\n","    k += 1\n","    if k > 15:\n","      #continue # Ensures that an infinite loop (where Tm and GC clamp requirements cannot be matched, e.g. if A/T rich region) is broken\n","      break\n","  fwd_primer_gc = gc_fraction(fwd_primer) * 100 # Calculate the GC content for the primer\n","\n","  fwd_primers[mutant + '_fwd'] = {'bases':fwd_primer, 'tm':round(fwd_primer_tm, 1), 'gc':round(fwd_primer_gc, 1)} # Store the nucleotides, Tm and GC content for the generated Fwd primer rounded to 1 decimal point1\n","\n","  # Generate the complementary Rev primer based on the selected Fwd primer\n","  n = 1\n","  fwd_primer_start = plasmid.find(fwd_primer[3:]) # Identifies the plasmid index where the Fwd primer starts\n","  rev_primer = (plasmid[fwd_primer_start - 35:fwd_primer_start - 3]) # Generates a Rev primer starting 30 bp upstream of the Fwd primer start\n","  #killing while loop Rv\n","  #while 'G' not in rev_primer[0] and 'C' not in rev_primer[0]:\n","  #  rev_primer = (plasmid[fwd_primer_start - 35 - n:fwd_primer_start - 3]) # Lengthens the Rev primer by one bp at a time if there is no GC clamp at the 3' position\n","  #  n += 1\n","  rev_primer_tm = mt.Tm_NN(rev_primer, nn_table=mt.DNA_NN3)\n","\n","  # Checks whether the Rev primer Tm is too high\n","  n = 1\n","  while rev_primer_tm > upper_tm:\n","    rev_primer = plasmid[fwd_primer_start - 35 + n:fwd_primer_start - 3] # Shortens the Rev primer by one bp at a time if the Tm is too high\n","    n += 1\n","    #if 'G' in rev_primer[0] or 'C' in rev_primer[0]:\n","    rev_primer_tm = mt.Tm_NN((rev_primer), nn_table=mt.DNA_NN3)\n","\n","  n = 1\n","  while rev_primer_tm < lower_tm:\n","    rev_primer = plasmid[fwd_primer_start - 35 - n:fwd_primer_start - 3] # Lengthens the Rev primer by one bp at a time if the Tm is too low\n","    n += 1\n","    #if 'G' in rev_primer[0] or 'C' in rev_primer[0]:\n","    rev_primer_tm = mt.Tm_NN((rev_primer), nn_table=mt.DNA_NN3)\n","\n","  rev_primer_rev_comp = str(Seq(rev_primer).reverse_complement())\n","\n","  rev_primer_gc = gc_fraction(rev_primer) * 100\n","\n","  rev_primers[mutant + '_rev'] = {'bases':rev_primer_rev_comp, 'tm':round(rev_primer_tm, 1), 'gc':round(rev_primer_gc, 1)} # Store the nucleotides, Tm and GC content for the generated Rev primer rounded to 1 decimal point"],"metadata":{"id":"sTECgdjJBnfJ","executionInfo":{"status":"ok","timestamp":1701362175201,"user_tz":0,"elapsed":907,"user":{"displayName":"Henry Taunt","userId":"07657040803472197154"}}},"execution_count":8,"outputs":[]},{"cell_type":"code","source":["df_fwd_primers = pd.DataFrame.from_dict(fwd_primers).transpose() # Convert the Fwd primer dictionary to a dataframe\n","df_rev_primers = pd.DataFrame.from_dict(rev_primers).transpose() # Convert the Rev primer dictionary to a dataframe\n","\n","df_primer_output = pd.DataFrame(columns = [''])\n","\n","df_primer_output = pd.concat([df_fwd_primers, df_rev_primers]) # Concatenate the Fwd and Rev primer dataframes"],"metadata":{"id":"dZtenEP1m-iJ","executionInfo":{"status":"ok","timestamp":1701362175201,"user_tz":0,"elapsed":5,"user":{"displayName":"Henry Taunt","userId":"07657040803472197154"}}},"execution_count":9,"outputs":[]},{"cell_type":"code","source":["# Change the output path to a common folder?\n","\n","df_primer_output.to_excel('/content/drive/My Drive/primer_output_' + lib_name + '.xlsx')"],"metadata":{"id":"Kk0KquDIKOrn","executionInfo":{"status":"ok","timestamp":1701362175596,"user_tz":0,"elapsed":398,"user":{"displayName":"Henry Taunt","userId":"07657040803472197154"}}},"execution_count":10,"outputs":[]}]}